src/wcl/filepath.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+
+namespace wcl {
+
+class filepath_iterator {
+ private:
+  // TODO: This should be a string_view
+  const std::string& str;
+  size_t start = 0, final = 0;
+
+  // This moves `final` to the next '/' or to the end of the path
+  // and moves `start` past '/' if its currently on it.
+  void next() {
+    if (final < str.size() && str[final] == '/') final++;
+    start = final;
+    while (final < str.size() && str[final] != '/') {
+      final++;
+    }
+  }
+
+ public:
+  explicit filepath_iterator(const std::string& str) : str(str), start(), final() { next(); }
```

Comment:
in that case, do you even need `start(), final()` in the initializer list? Their presence implies that they are doing something but I don't think they are

Reply:
Yeah but only because I have ' = 0' added above, otherwise they would be unset. Primitives are treated ever so subtly differently in C++ in this respect.

I try to add them to make sure its all initialized, I guess I'm paranoid.